OrbitVulkanLayer/SubmissionTrackerTest.cpp

```diff
@@ -0,0 +1,1446 @@
+
+  tracker_.TrackCommandBuffers(device_, command_pool_, &command_buffer_, 1);
+  tracker_.MarkCommandBufferBegin(command_buffer_);
+}
+
+TEST_F(SubmissionTrackerTest, MarkCommandBufferBeginWillWriteTimestampWhenCapturing) {
+  // We cannot capture anything in that lambda as we need to cast it to a function pointer.
+  static bool was_called = false;
+
+  PFN_vkCmdWriteTimestamp mock_write_timestamp_function =
+      +[](VkCommandBuffer /*command_buffer*/, VkPipelineStageFlagBits /*pipeline_stage*/,
+          VkQueryPool /*query_pool*/, uint32_t query) {
+        EXPECT_EQ(query, kSlotIndex1);
+        was_called = true;
+      };
+
+  EXPECT_CALL(timer_query_pool_, NextReadyQuerySlot)
+      .Times(1)
+      .WillOnce(Invoke(MockNextReadyQuerySlot1));
+  EXPECT_CALL(dispatch_table_, CmdWriteTimestamp)
+      .Times(1)
+      .WillOnce(Return(mock_write_timestamp_function));
+
+  producer_->StartCapture();
+  tracker_.TrackCommandBuffers(device_, command_pool_, &command_buffer_, 1);
+  tracker_.MarkCommandBufferBegin(command_buffer_);
+
+  EXPECT_TRUE(was_called);
+  was_called = false;
+}
+
+TEST_F(SubmissionTrackerTest, ResetCommandBufferShouldRollbackUnsubmittedSlots) {
+  EXPECT_CALL(timer_query_pool_, NextReadyQuerySlot)
+      .Times(1)
+      .WillOnce(Invoke(MockNextReadyQuerySlot1));
+  std::vector<uint32_t> actual_slots_to_rollback;
+  EXPECT_CALL(timer_query_pool_, RollbackPendingQuerySlots)
+      .Times(1)
+      .WillOnce(SaveArg<1>(&actual_slots_to_rollback));
+
+  producer_->StartCapture();
+  tracker_.TrackCommandBuffers(device_, command_pool_, &command_buffer_, 1);
+  tracker_.MarkCommandBufferBegin(command_buffer_);
+  tracker_.ResetCommandBuffer(command_buffer_);
+
+  EXPECT_THAT(actual_slots_to_rollback, ElementsAre(kSlotIndex1));
+}
+
+TEST_F(SubmissionTrackerTest, ResetCommandPoolShouldRollbackUnsubmittedSlots) {
+  EXPECT_CALL(timer_query_pool_, NextReadyQuerySlot)
+      .Times(1)
+      .WillOnce(Invoke(MockNextReadyQuerySlot1));
+  std::vector<uint32_t> actual_slots_to_rollback;
+  EXPECT_CALL(timer_query_pool_, RollbackPendingQuerySlots)
+      .Times(1)
+      .WillOnce(SaveArg<1>(&actual_slots_to_rollback));
+
+  producer_->StartCapture();
+  tracker_.TrackCommandBuffers(device_, command_pool_, &command_buffer_, 1);
+  tracker_.MarkCommandBufferBegin(command_buffer_);
+  tracker_.ResetCommandPool(command_pool_);
+
+  EXPECT_THAT(actual_slots_to_rollback, ElementsAre(kSlotIndex1));
+}
+
+TEST_F(SubmissionTrackerTest, MarkCommandBufferEndWontWriteTimestampsWhenNotCapturing) {
+  EXPECT_CALL(timer_query_pool_, NextReadyQuerySlot).Times(0);
+  EXPECT_CALL(dispatch_table_, CmdWriteTimestamp).Times(0);
+
+  tracker_.TrackCommandBuffers(device_, command_pool_, &command_buffer_, 1);
+  tracker_.MarkCommandBufferBegin(command_buffer_);
+  tracker_.MarkCommandBufferEnd(command_buffer_);
+}
+
+TEST_F(SubmissionTrackerTest, MarkCommandBufferEndWillWriteTimestampsWhenNotCapturedBegin) {
```

Comment:
I would change the end to "when begin not captured".

Reply:
Done.